### PR TITLE
Beacon API: update GetStateRoot 

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

Address feedbacks from #8402

**Which issues(s) does this PR fix?**

N/A

**Other notes for review**

N/A

### CODE REVIEW

File: beacon-chain/rpc/beaconv1/state.go [hunk 1]
```diff
@@ -159,21 +167,23 @@ func (bs *Server) stateRootByHex(ctx context.Context, stateId []byte) ([]byte, e
...
-		return nil, status.Errorf(codes.Internal, "Failed to obtain state: %v", err)
+		return nil, status.Errorf(codes.Internal, "Could not get blocks: %v", err)
 	}
-	stateRoot, err := state.HashTreeRoot(ctx)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	if !e {
+		return nil, status.Errorf(codes.NotFound, "No block exists")
+	}
+	if len(blks) != 1 {
```
Comment (terencechain) [inline beacon-chain/rpc/beaconv1/state.go:185]: Note: I checked eth2api and I'm not sure what to do here. There could be multiple blocks exist in the same slot. In that case, which state root do we use?

**Comment by terencechain** [comment]
cc @0xKiwi @rkapka

File: beacon-chain/rpc/beaconv1/state_test.go [hunk 2]
```diff
@@ -76,95 +76,90 @@ func TestGetGenesis(t *testing.T) {
...
-			ChainInfoFetcher: chainService,
+			ChainInfoFetcher: &chainMock.ChainService{Block: b},
 		}
+
 		resp, err := s.GetStateRoot(ctx, &ethpb.StateRequest{
 			StateId: []byte("head"),
 		})
 		require.NoError(t, err)
-		assert.DeepEqual(t, root, resp.Data.StateRoot)
+		assert.DeepEqual(t, resp.Data.StateRoot, resp.Data.StateRoot)
```
Comment (rkapka) [inline beacon-chain/rpc/beaconv1/state_test.go:93]: You are comparing the same things here.
```suggestion
		assert.DeepEqual(t, bytesutil.PadTo([]byte("head"), 32), resp.Data.StateRoot)
```

File: beacon-chain/rpc/beaconv1/state.go (hunk 1 above)
Comment (rkapka) [reply beacon-chain/rpc/beaconv1/state.go:185]: Maybe ask this question on the APIs channel? https://discord.com/channels/595666850260713488/710831784991916072

File: beacon-chain/rpc/beaconv1/state_test.go [hunk 3]
```diff
@@ -217,45 +212,54 @@ func TestGetStateRoot(t *testing.T) {
...
 		resp, err := s.GetStateRoot(ctx, &ethpb.StateRequest{
 			StateId: []byte("100"),
 		})
 		require.NoError(t, err)
-		expectedRoot, err := state.HashTreeRoot(ctx)
-		require.NoError(t, err)
-		var b [32]byte
-		copy(b[:], resp.Data.StateRoot)
-		assert.DeepEqual(t, expectedRoot, b)
+		assert.DeepEqual(t, resp.Data.StateRoot, bytesutil.PadTo([]byte("slot"), 32))
```
Comment (rkapka) [inline beacon-chain/rpc/beaconv1/state_test.go:232]:
```suggestion
		assert.DeepEqual(t, bytesutil.PadTo([]byte("slot"), 32), resp.Data.StateRoot)
```

File: beacon-chain/rpc/beaconv1/state_test.go [hunk 4]
```diff
@@ -76,95 +76,90 @@ func TestGetGenesis(t *testing.T) {
...
 		resp, err := s.GetStateRoot(ctx, &ethpb.StateRequest{
 			StateId: []byte("justified"),
 		})
 		require.NoError(t, err)
-		expectedRoot, err := state.HashTreeRoot(ctx)
-		require.NoError(t, err)
-		var b [32]byte
-		copy(b[:], resp.Data.StateRoot)
-		assert.DeepEqual(t, expectedRoot, b)
+		assert.DeepEqual(t, resp.Data.StateRoot, bytesutil.PadTo([]byte("justified"), 32))
```
Comment (rkapka) [inline beacon-chain/rpc/beaconv1/state_test.go:162]:
```suggestion
		assert.DeepEqual(t, bytesutil.PadTo([]byte("justified"), 32), resp.Data.StateRoot)
```

File: beacon-chain/rpc/beaconv1/state_test.go [hunk 5]
```diff
@@ -76,95 +76,90 @@ func TestGetGenesis(t *testing.T) {
...
 		resp, err := s.GetStateRoot(ctx, &ethpb.StateRequest{
 			StateId: []byte("finalized"),
 		})
 		require.NoError(t, err)
-		expectedRoot, err := state.HashTreeRoot(ctx)
-		require.NoError(t, err)
-		var b [32]byte
-		copy(b[:], resp.Data.StateRoot)
-		assert.DeepEqual(t, expectedRoot, b)
+		assert.DeepEqual(t, resp.Data.StateRoot, bytesutil.PadTo([]byte("finalized"), 32))
```
Comment (rkapka) [inline beacon-chain/rpc/beaconv1/state_test.go:137]:
```suggestion
		assert.DeepEqual(t, bytesutil.PadTo([]byte("finalized"), 32), resp.Data.StateRoot)
```

File: beacon-chain/rpc/beaconv1/state_test.go [hunk 6]
```diff
@@ -76,95 +76,90 @@ func TestGetGenesis(t *testing.T) {
...
 		resp, err := s.GetStateRoot(ctx, &ethpb.StateRequest{
 			StateId: []byte("genesis"),
 		})
 		require.NoError(t, err)
-		expectedRoot, err := state.HashTreeRoot(ctx)
-		require.NoError(t, err)
-		var b [32]byte
-		copy(b[:], resp.Data.StateRoot)
-		assert.DeepEqual(t, expectedRoot, b)
+		assert.DeepEqual(t, resp.Data.StateRoot, bytesutil.PadTo([]byte("genesis"), 32))
```
Comment (rkapka) [inline beacon-chain/rpc/beaconv1/state_test.go:112]:
```suggestion
		assert.DeepEqual(t, bytesutil.PadTo([]byte("genesis"), 32), resp.Data.StateRoot)
```

**Review by rkapka — CHANGES_REQUESTED** [review]
Requesting changes for the test where the same things are compared. A very minor thing is the incorrect order of expected/actual in other tests (it's annoying that it cannot be enforced...).

**Comment by terencechain** [comment]
> Maybe ask this question on the APIs channel?

Let's just leave it as it is now. I like the current approach as simpler. Anything fancier to be more of a dos issue

**Review by rkapka — APPROVED** [review]
Looks great, thanks for fixing my mess!

**Comment by terencechain** [comment]
> Looks great, thanks for fixing my mess!

Anytime. You already did great amount of work there, I just modified a few things :)

**Pull request merged**